lib/src/mock_query.dart

```diff
@@ -239,55 +251,182 @@ class MockQuery<T extends Object?> extends FakeQueryWithParent<T> {
       Iterable<Object?>? whereIn,
       Iterable<Object?>? whereNotIn,
       bool? isNull}) {
-    final operation = (List<DocumentSnapshot<T>> docs) =>
-        docs.where((document) {
-          dynamic value;
-          if (field is String || field is FieldPath) {
-            // DocumentSnapshot.get can throw StateError
-            // if field cannot be found. In query it does not matter,
-            // so catch and set value to null.
-            try {
-              value = document.get(field);
-            } on StateError catch (_) {
-              value = null;
-            }
-          } else if (field == FieldPath.documentId) {
-            value = document.id;
-
-            // transform any DocumentReference in the query to id.
-            isEqualTo = transformValue(isEqualTo, documentReferenceToId);
-            isNotEqualTo = transformValue(isNotEqualTo, documentReferenceToId);
-            isLessThan = transformValue(isLessThan, documentReferenceToId);
-            isLessThanOrEqualTo =
-                transformValue(isLessThanOrEqualTo, documentReferenceToId);
-            isGreaterThan =
-                transformValue(isGreaterThan, documentReferenceToId);
-            isGreaterThanOrEqualTo =
-                transformValue(isGreaterThanOrEqualTo, documentReferenceToId);
-            arrayContains =
-                transformValue(arrayContains, documentReferenceToId);
-            arrayContainsAny =
-                transformValue(arrayContainsAny, documentReferenceToId);
-            whereIn = transformValue(whereIn, documentReferenceToId);
-            whereNotIn = transformValue(whereNotIn, documentReferenceToId);
-            isNull = transformValue(isNull, documentReferenceToId);
-          }
-          return _valueMatchesQuery(value,
-              isEqualTo: isEqualTo,
-              isNotEqualTo: isNotEqualTo,
-              isLessThan: isLessThan,
-              isLessThanOrEqualTo: isLessThanOrEqualTo,
-              isGreaterThan: isGreaterThan,
-              isGreaterThanOrEqualTo: isGreaterThanOrEqualTo,
-              arrayContains: arrayContains,
-              arrayContainsAny: arrayContainsAny,
-              whereIn: whereIn,
-              whereNotIn: whereNotIn,
-              isNull: isNull);
-        }).toList();
+    if (field is Filter) {
```

Comment:
At first glance, it is a bit difficult to see that the only thing that changes is the predicate. We could rewrite it like so:
```dart
final predicate = field is Filter ?
  _buildFilterPredicate(field.toJson())
  : (document) => _wherePredicate(document, field,
      isEqualTo: isEqualTo,
      isNotEqualTo: isNotEqualTo,
      isLessThan: isLessThan,
      isLessThanOrEqualTo: isLessThanOrEqualTo,
      isGreaterThan: isGreaterThan,
      isGreaterThanOrEqualTo: isGreaterThanOrEqualTo,
      arrayContains: arrayContains,
      arrayContainsAny: arrayContainsAny,
      whereIn: whereIn,
      whereNotIn: whereNotIn,
      isNull: isNull)
final operation = (List<DocumentSnapshot<T>> docs) => docs
  .where(predicate)
  .toList();
return MockQuery<T>(this, operation);
```

Reply:
I made the change at https://github.com/atn832/fake_cloud_firestore/commit/9af8832de35a8ff810a08210a8fded7b741bae83.